src/rust/metric-forwarder/src/accumulate_metrics.rs

```diff
@@ -0,0 +1,250 @@
+use chrono::{DateTime, Duration, DurationRound, FixedOffset};
+
+use ordered_float::OrderedFloat;
+
+use rusoto_cloudwatch::{Dimension, MetricDatum};
+
+use std::collections::HashMap;
+use std::hash::{Hash, Hasher};
+
+type CountMap = HashMap<OrderedFloat<f64>, f64>;
+type DatumToCountMap = HashMap<WrappedMetricDatum, CountMap>;
+
+pub fn accumulate_metric_data(input: Vec<MetricDatum>) -> Vec<MetricDatum> {
+    let mut datum_to_count: DatumToCountMap = HashMap::new();
+    /*
+     * This section of the code will convert input of (boiled down for simplicity)
+     *   [MetricDatum('metric', 1.0), MetricDatum('metric', 5.0), MetricDatum('metric', 1.0)]
+     * into
+     * {
+     *   MetricDatum('metric', _): {
+     *     // value -> num_occurences
+     *     1.0: 2,
+     *     5.0: 1
+     *   }
+     * }
+     */
+    input.into_iter().for_each(|datum| {
+        let value = &datum.value.unwrap();
+        let count_map = datum_to_count
+            .entry(WrappedMetricDatum(datum))
+            .or_insert(HashMap::new());
+        let count = count_map.entry(OrderedFloat(*value)).or_insert(0f64);
+        (*count) += 1f64;
+    });
+
+    /*
+     * This section will take the above `datum_to_count_map` and convert it into a flat vec<MetricDatum>, where each
+     * MetricDatum internally contains the {1.0: 2, 5.0: 1} in its .counts and .values
+     */
+    datum_to_count
+        .into_iter()
+        .map(|(WrappedMetricDatum(mut datum), count_map)| {
+            // Erase the scalar "value", then fill "counts" and "values" vecs
+            datum.value = None;
+            let mut counts = vec![];
+            let mut values: Vec<f64> = vec![];
```

Comment:
I think you could do *something* like
```rust
let (counts, values) = count_map.drain().partition();
```
Or at the least it seems like a fine place to do a `Vec::with_capacity(count_map.len())`?

Reply:
partition() doesn't work for this use case afaict, since it would be deciding "okay, given a specific entry, like (k,v), does it go into FirstCollection or SecondCollection?" 

basically, what I want is to extract all the keys and all the values, in the same order.